docs/apps/hyperqueue.md

```diff
@@ -4,234 +4,280 @@ tags:
-official documentation.
+**Server**
 
-Next, we can start submitting jobs or alternatively wait for all the workers to connect
-to the server before submission. This is generally good practice as we can notice issues with
-the workers early.
-
-Wait until all workers are online (note no timeout):
+Next, we specify where HyperQueue places the server files.
+All `hq` commands respect this variable, so we set it before using any `hq` commands.
+If a server directory is not specified, it will default to the user's home directory.
+In this case, one has to be careful not to mix up separate computations. We recommend starting one server per job in a job-specific directory for simple cases that fit inside one Slurm job.
 
 ```bash
-hq worker wait "${SLURM_NTASKS}"
-```
-
-### Submitting jobs
+# Specify a location for the server
+export HQ_SERVER_DIR="$PWD/.hq-server/$SLURM_JOB_ID"
 
-!!!info "Output"
-    By default HQ creates one folder for each job where output is redirected.
-    You can use the `--log`, `--stdout` and `--stderr` flags to change this behavior.
-    Note that it's not possible to direct output from multiple jobs into the same file
-    as each submission will clear the file.
+# Create a directory for the server
+mkdir -p "$HQ_SERVER_DIR"
+```
 
-See `hq submit --help` for the full list of options.
+Now, we start the server in the background and wait for it to start.
+The server keeps running until we stop it; therefore, we place it in the background so it does not block the execution of the rest of the script.
 
 ```bash
-hq submit <hq submit args> --cpus <n> <COMMAND/executable> <args to program>
+# Start the server in the background
+hq server start &
+
+# Wait until the server has started
+until hq job list &> /dev/null ; do sleep 1 ; done
 ```
 
-This is a non-blocking command similar to `sbatch`.
+**Workers**
 
-HyperQueue is not limited to running a single execution per submission. Using the
-`--array 1-N` flag we can start a program *N* times similar to how Slurm array jobs work.
+Next, we start HyperQueue workers in the background with the number of CPUs and the amount of memory defined in the batch script.
+We access those values using the `SLURM_CPU_PER_TASK` and `SLURM_MEM_PER_CPU` environment variables.
+By starting the workers using the `srun` command, we create one worker per Slurm task.
+We also wait for all workers to connect, which is generally good practice as we can notice issues with the workers early.
 
 ```bash
-hq submit --array 1-10 --cpus <n> <COMMAND>
+# Set memory for workers in bytes according to SLURM_MEM_PER_CPU if greater than zero.
+# Otherwise, leave unset to use all the memory of the node.
+if [[ "${SLURM_MEM_PER_CPU:-0}" -gt 0 ]]; then
+    # Calculate the total memory reservation and convert it from megabytes to bytes.
+    TOTAL_MEM_BYTES=$((SLURM_CPUS_PER_TASK * SLURM_MEM_PER_CPU * 1000000))
+    TOTAL_MEM_OPT="--resource mem=sum($TOTAL_MEM_BYTES)"
+else
+    TOTAL_MEM_OPT=""
+fi
```

Comment:
Is this really necessary? I've used hyperqueue without this and haven't noticed memory issues.

Reply:
This code is indeed quite ugly.
If HyperQueue works well without this, we should remove it.
I suppose it is up to the programmer to not exceed the memory limit.

Reply:
I'd propose to remove this for now in order not to make the example to complicated.